src/source/graphite.rs

```diff
@@ -64,87 +68,165 @@ impl Graphite {
+                    poller
+                        .register(
+                            &stream,
+                            mio::Token(0),
+                            mio::Ready::readable(),
+                            mio::PollOpt::edge(),
+                        )
+                        .unwrap();
+
+                    handle_stream(rchans, rtags, poller, stream);
                 });
+                stream_handlers.push(new_stream);
             }
-        }
-    })
-}
 
+            Err(e) => match e.kind() {
+                std::io::ErrorKind::WouldBlock => {
+                    break;
+                }
+                _ => unimplemented!(),
+            },
+        };
+    }
+}
 
 fn handle_stream(
     mut chans: util::Channel,
-    tags: Arc<metric::TagMap>,
-    stream: TcpStream,
+    tags: sync::Arc<metric::TagMap>,
+    poller: mio::Poll,
+    stream: mio::net::TcpStream,
+) {
+    let mut line = String::new();
+    let mut res = Vec::new();
+    let mut line_reader = BufReader::new(stream);
+    let basic_metric = sync::Arc::new(Some(
+        metric::Telemetry::default().overlay_tags_from_map(&tags),
+    ));
+
+    loop {
+        let mut events = mio::Events::with_capacity(1024);
+        match poller.poll(&mut events, None) {
+            Err(e) => panic!(format!("Failed during poll {:?}", e)),
+            Ok(_num_events) => for event in events {
+                match event.token() {
+                    constants::SYSTEM => return,
+                    _stream_token => if let Ok(len) = line_reader.read_line(&mut line)
+                    {
+                        if len > 0 {
+                            if parse_graphite(&line, &mut res, &basic_metric) {
+                                assert!(!res.is_empty());
+                                GRAPHITE_GOOD_PACKET.fetch_add(1, Ordering::Relaxed);
+                                GRAPHITE_TELEM.fetch_add(1, Ordering::Relaxed);
+                                for m in res.drain(..) {
+                                    send(
+                                        &mut chans,
+                                        metric::Event::Telemetry(
+                                            sync::Arc::new(Some(m)),
+                                        ),
+                                    );
+                                }
+                                line.clear();
+                            } else {
+                                GRAPHITE_BAD_PACKET.fetch_add(1, Ordering::Relaxed);
+                                error!("bad packet: {:?}", line);
+                                line.clear();
+                            }
+                        } else {
+                            break;
+                        }
+                    },
+                }
+            },
+        }
+    }
+}
+
+fn handle_tcp(
+    chans: util::Channel,
+    tags: sync::Arc<metric::TagMap>,
+    socket_map: HashMap<mio::Token, mio::net::TcpListener>,
```

Comment:
Oops, `HashMap` in the wild. Hoping to convert this to slab.

Reply:
Fixed.